<!DOCTYPE html SYSTEM "about:legacy-compat">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<META http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>fastqscreen </title>
<style type="text/css">

    #genomeSample{
        color:#9B1319;
        font-style : bold;
    }

    tr.sampleDataTab:hover {
      z-index:2;
      box-shadow:0 0 12px rgba(0, 0, 0, 1);
      background:#F6F6B4;
    }

    td.sampleDataTab {
        width:90px
    }

    td {
      text-align: center;
      width: 100px;
      border: 1px solid black;

      padding-left:3px;
      padding-right:3px;
      padding-top:1px;
      padding-bottom:1px;
      position : static;
      background-clip: padding-box;
    }
    th {
       background-color:#E7EAF1;
       color:black;
       border: thin solid black;
       border-bottom-width: 2px;
       width: 50px;
       font-size: 100%;
    }
    table {
       border: medium solid #000000;
       font-size: 95%;
       border-collapse:collapse;
    }
    img {
        width: 150px;
    }
    body{
       font-family: sans-serif;
       font-size: 80%;
       margin:0;
    }
    h1{
       color: #234CA5;
       font-style : italic;
       font-size : 20px;
    }
    h2{}
    h3{
       color:black;
       font-style : italic;
    }
    div.header {
        background-color: #C1C4CA;
        border:0;
        margin:0;
        padding: 0.5em;
        font-size: 175%;
        font-weight: bold;
        width:100%;
        height: 2em;
        position: fixed;
        vertical-align: middle;
        z-index:2;

    }
    #header_title {
        display:inline-block;
        float:left;
        clear:left;
    }
    div.report {
        display:block;
        position:absolute;
        width:100%;
        top:5.2em;
        bottom:5px;
        left:0;
        right:0;
        padding:0 1em 0 1em;
        background-color: white;
    }
    div.projectReport {
        display:block;
        position:absolute;
        width:80%;
        top:18em;
        bottom:5px;
        left:0;
        right:0;
        padding:0 0 0 1em;
        background-color: white;
    }

    div.footer {
        background-color: #C1C4CA;
        border:0;
        margin:0;
        padding:0.5em;
        height: 1.3em;
        overflow:hidden;
        font-size: 100%;
        font-weight: bold;
        position:fixed;
        bottom:0;
        width:100%;
        z-index:2;
   }

    .report_left{
      display:inline-block;
      float:left;
      clear:left;
      text-align:left;
      vertical-align:middle;
      width:auto;
      width:60%;
      padding-bottom:3em;
    }

    .report_right{
      display:inline-block;
      float:right;
      clear:right;
      text-align:left;
      vertical-align:middle;
      width:auto;
      width:40%;
      padding-bottom:3em;
    }


    .linkSample {
    border-top-left-radius:24px;
    border-top-right-radius:24px;
    border-bottom-right-radius:24px;
    border-bottom-left-radius:24px;

    background-color:#ffffff;
    text-indent:0px;
    border:3px solid #000000;
    display:inline-block;
    font-family:Arial;
    font-size:12px;
    font-weight:bold;
    font-style:normal;
    line-height:29px;
    text-decoration:none;
    text-align:center;
    padding-left:5px;
    padding-right:5px;
    white-space:nowrap;
    }

    #goToSample td{
    border:#ffffff;
    border-style:hidden;
    }

    div.runReport{
    border:0;
    clear:inherit;
    margin:0;
    width:100%;
    height: 17em;
    background-color: white;
    position: fixed;
    z-index:2;
    }
  </style>
<script type="text/javascript">
<!--

        function goto(){
            // Decallage de 2 pour corriger l'affichage
            var index = document.forms.sampleName.nom.options.selectedIndex - 2;
            var anchor = document.forms.sampleName.nom.options[index].text;

            window.location.href = "#"+anchor;
        }

    --></script>
</head>
<body>
<div class="header">
<div id="header_title">
<img src="http://outils.genomique.biologie.ens.fr/aozan/images/logo_aozan_qc.png" alt="Aozan">
        Detection contamination report for project <b>Project_name_1</b>
</div>
</div>
<a name="top"></a>
<div class="report">
<div class="runReport">
<ul>
<li>
<b>Run Id: </b>141009_SNL110_0139_AC5GMMACXX</li>
<li>
<b>Flow cell: </b>C5GMMACXX</li>
<li>
<b>Date: </b>Thu 09 Oct 2014</li>
<li>
<b>Instrument S/N: </b>SNL110</li>
<li>
<b>Instrument run number: </b>139</li>
<li>
<b>Generated by: </b>Aozan version 1.4</li>
<li>
<b>Creation date: </b>Mon 16 Feb 2015</li>
<li>
<b>Project : </b>Project_name_1</li>
<li>
<b>Genome : </b>hg19</li>
</ul>
<form name="sampleName">
<label>Display a sample : </label><select name="nom" size="1" onchange="goto()"><option>2014_437</option><option>2014_439</option><option>2014_440</option><option>2014_441</option><option>2014_442</option></select>
</form>
</div>
<div class="projectReport">
<div class="report_left">
<h3>Sample: 2014_437</h3>
<a id="2014_437"></a>
<table border="1" class="sampleDataTab">
<tr>
<th>Library</th><th>%Mapped</th><th>%Unmapped</th><th>%One hit one library</th><th>%Multiple hits one library</th><th>%One hit multiple libraries</th><th>%Multiple hits multiple libraries</th>
</tr>
<tr>
<td>adapters</td><td>1.50 %</td><td>98.50 %</td><td>1.45 %</td><td>0.05 %</td><td>0.00 %</td><td>0.00 %</td>
</tr>
<tr>
<td>lsuref</td><td>0.08 %</td><td>99.92 %</td><td>0.00 %</td><td>0.00 %</td><td>0.01 %</td><td>0.06 %</td>
</tr>
<tr>
<td>ssuref</td><td>0.08 %</td><td>99.92 %</td><td>0.00 %</td><td>0.00 %</td><td>0.02 %</td><td>0.06 %</td>
</tr>
<tr id="genomeSample">
<td>hg19</td><td>72.33 %</td><td>27.67 %</td><td>58.58 %</td><td>13.65 %</td><td>0.00 %</td><td>0.10 %</td>
</tr>
<tr>
<td>phix</td><td>0.91 %</td><td>99.09 %</td><td>0.91 %</td><td>0.00 %</td><td>0.00 %</td><td>0.00 %</td>
</tr>
</table>
</div>
<div class="report_right">
<ul>
<li>
<b>Genome : </b>hg19</li>
<li>
<b>Description: </b>
</li>
</ul>
<p>
<ul>
<li>reads_unmapped_none_genome : 25.25 %</li>
<li>reads_mapped_at_least_one_genome : 74.75 %</li>
<li>reads_mapped_at_except_genome_sample : 2.52 %</li>
<li>149 493&nbsp;
       reads_mapped  /
       200 000&nbsp;
       reads_processed</li>
</ul>
</p>
</div>
<div class="report_left">
<h3>Sample: 2014_439</h3>
<a id="2014_439"></a>
<table border="1" class="sampleDataTab">
<tr>
<th>Library</th><th>%Mapped</th><th>%Unmapped</th><th>%One hit one library</th><th>%Multiple hits one library</th><th>%One hit multiple libraries</th><th>%Multiple hits multiple libraries</th>
</tr>
<tr>
<td>adapters</td><td>0.62 %</td><td>99.38 %</td><td>0.51 %</td><td>0.10 %</td><td>0.00 %</td><td>0.00 %</td>
</tr>
<tr>
<td>lsuref</td><td>0.11 %</td><td>99.89 %</td><td>0.00 %</td><td>0.00 %</td><td>0.01 %</td><td>0.10 %</td>
</tr>
<tr>
<td>ssuref</td><td>0.11 %</td><td>99.89 %</td><td>0.00 %</td><td>0.00 %</td><td>0.02 %</td><td>0.09 %</td>
</tr>
<tr id="genomeSample">
<td>hg19</td><td>85.26 %</td><td>14.74 %</td><td>70.01 %</td><td>15.11 %</td><td>0.00 %</td><td>0.14 %</td>
</tr>
<tr>
<td>phix</td><td>1.66 %</td><td>98.34 %</td><td>1.66 %</td><td>0.00 %</td><td>0.00 %</td><td>0.00 %</td>
</tr>
</table>
</div>
<div class="report_right">
<ul>
<li>
<b>Genome : </b>hg19</li>
<li>
<b>Description: </b>
</li>
</ul>
<p>
<ul>
<li>reads_unmapped_none_genome : 12.46 %</li>
<li>reads_mapped_at_least_one_genome : 87.54 %</li>
<li>reads_mapped_at_except_genome_sample : 2.42 %</li>
<li>175 085&nbsp;
       reads_mapped  /
       200 000&nbsp;
       reads_processed</li>
</ul>
</p>
</div>
<div class="report_left">
<h3>Sample: 2014_440</h3>
<a id="2014_440"></a>
<table border="1" class="sampleDataTab">
<tr>
<th>Library</th><th>%Mapped</th><th>%Unmapped</th><th>%One hit one library</th><th>%Multiple hits one library</th><th>%One hit multiple libraries</th><th>%Multiple hits multiple libraries</th>
</tr>
<tr>
<td>adapters</td><td>0.44 %</td><td>99.56 %</td><td>0.39 %</td><td>0.04 %</td><td>0.00 %</td><td>0.00 %</td>
</tr>
<tr>
<td>lsuref</td><td>0.11 %</td><td>99.89 %</td><td>0.00 %</td><td>0.00 %</td><td>0.02 %</td><td>0.09 %</td>
</tr>
<tr>
<td>ssuref</td><td>0.10 %</td><td>99.90 %</td><td>0.00 %</td><td>0.00 %</td><td>0.02 %</td><td>0.08 %</td>
</tr>
<tr id="genomeSample">
<td>hg19</td><td>91.12 %</td><td>8.88 %</td><td>73.74 %</td><td>17.24 %</td><td>0.00 %</td><td>0.13 %</td>
</tr>
<tr>
<td>phix</td><td>0.45 %</td><td>99.55 %</td><td>0.44 %</td><td>0.00 %</td><td>0.00 %</td><td>0.00 %</td>
</tr>
</table>
</div>
<div class="report_right">
<ul>
<li>
<b>Genome : </b>hg19</li>
<li>
<b>Description: </b>
</li>
</ul>
<p>
<ul>
<li>reads_unmapped_none_genome : 8.00 %</li>
<li>reads_mapped_at_least_one_genome : 92.00 %</li>
<li>reads_mapped_at_except_genome_sample : 1.02 %</li>
<li>183 999&nbsp;
       reads_mapped  /
       200 000&nbsp;
       reads_processed</li>
</ul>
</p>
</div>
<div class="report_left">
<h3>Sample: 2014_441</h3>
<a id="2014_441"></a>
<table border="1" class="sampleDataTab">
<tr>
<th>Library</th><th>%Mapped</th><th>%Unmapped</th><th>%One hit one library</th><th>%Multiple hits one library</th><th>%One hit multiple libraries</th><th>%Multiple hits multiple libraries</th>
</tr>
<tr>
<td>adapters</td><td>2.20 %</td><td>97.80 %</td><td>2.12 %</td><td>0.08 %</td><td>0.00 %</td><td>0.00 %</td>
</tr>
<tr>
<td>lsuref</td><td>0.08 %</td><td>99.92 %</td><td>0.00 %</td><td>0.00 %</td><td>0.01 %</td><td>0.06 %</td>
</tr>
<tr>
<td>ssuref</td><td>0.08 %</td><td>99.92 %</td><td>0.00 %</td><td>0.00 %</td><td>0.02 %</td><td>0.06 %</td>
</tr>
<tr id="genomeSample">
<td>hg19</td><td>67.52 %</td><td>32.48 %</td><td>54.58 %</td><td>12.83 %</td><td>0.00 %</td><td>0.10 %</td>
</tr>
<tr>
<td>phix</td><td>0.37 %</td><td>99.63 %</td><td>0.36 %</td><td>0.00 %</td><td>0.00 %</td><td>0.00 %</td>
</tr>
</table>
</div>
<div class="report_right">
<ul>
<li>
<b>Genome : </b>hg19</li>
<li>
<b>Description: </b>
</li>
</ul>
<p>
<ul>
<li>reads_unmapped_none_genome : 29.91 %</li>
<li>reads_mapped_at_least_one_genome : 70.09 %</li>
<li>reads_mapped_at_except_genome_sample : 2.67 %</li>
<li>140 182&nbsp;
       reads_mapped  /
       200 000&nbsp;
       reads_processed</li>
</ul>
</p>
</div>
<div class="report_left">
<h3>Sample: 2014_442</h3>
<a id="2014_442"></a>
<table border="1" class="sampleDataTab">
<tr>
<th>Library</th><th>%Mapped</th><th>%Unmapped</th><th>%One hit one library</th><th>%Multiple hits one library</th><th>%One hit multiple libraries</th><th>%Multiple hits multiple libraries</th>
</tr>
<tr>
<td>adapters</td><td>0.67 %</td><td>99.33 %</td><td>0.64 %</td><td>0.03 %</td><td>0.00 %</td><td>0.00 %</td>
</tr>
<tr>
<td>lsuref</td><td>0.08 %</td><td>99.92 %</td><td>0.00 %</td><td>0.00 %</td><td>0.01 %</td><td>0.07 %</td>
</tr>
<tr>
<td>ssuref</td><td>0.08 %</td><td>99.92 %</td><td>0.00 %</td><td>0.00 %</td><td>0.01 %</td><td>0.07 %</td>
</tr>
<tr id="genomeSample">
<td>hg19</td><td>85.50 %</td><td>14.50 %</td><td>69.01 %</td><td>16.38 %</td><td>0.00 %</td><td>0.10 %</td>
</tr>
<tr>
<td>phix</td><td>0.56 %</td><td>99.44 %</td><td>0.56 %</td><td>0.00 %</td><td>0.00 %</td><td>0.00 %</td>
</tr>
</table>
</div>
<div class="report_right">
<ul>
<li>
<b>Genome : </b>hg19</li>
<li>
<b>Description: </b>
</li>
</ul>
<p>
<ul>
<li>reads_unmapped_none_genome : 13.27 %</li>
<li>reads_mapped_at_least_one_genome : 86.73 %</li>
<li>reads_mapped_at_except_genome_sample : 1.33 %</li>
<li>173 461&nbsp;
       reads_mapped  /
       200 000&nbsp;
       reads_processed</li>
</ul>
</p>
</div>
</div>
</div>
<div class="footer">
    Generated by
    <a href="http://transcriptome.ens.fr/aozan">Aozan</a>
    (version 1.4)

</div>
</body>
</html>
